src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java

```diff
@@ -1004,6 +1043,105 @@ public boolean wasModifiedSinceDigest(Path path) {
     }
   }
 
+  private static final class FileWriteOutputArtifactValue extends FileArtifactValue {
+    private final DeterministicWriter writer;
+    private final long size;
+    private final byte[] digest;
+    @Nullable private FileContentsProxy proxy;
+
+    private FileWriteOutputArtifactValue(DeterministicWriter writer, long size, byte[] digest) {
+      this.writer = writer;
+      this.size = size;
+      this.digest = digest;
+    }
+
+    @Override
+    public FileStateType getType() {
+      return FileStateType.REGULAR_FILE;
+    }
+
+    @Override
+    public byte[] getDigest() {
+      return digest;
+    }
+
+    @Override
+    public long getSize() {
+      return size;
+    }
+
+    @Override
+    public boolean isInline() {
+      return true;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+      // TODO: Avoid materializing the full content in memory by using a variant of
+      //  Piped{Input,Output}Stream that works well with virtual threads.
```

Comment:
@justinhorvitz Is this something that has come up for you before? It could be particularly useful for `VirtualActionInput`'s `StreamWriter` implementation in Bazel.

Reply:
Uploading virutal inputs for remote execution can cost a lot of memory, so internally we use a semaphore to limit the number of inflight bytes. For large inputs, we use a "split store" technique where fixed-length chunks of the file are uploaded in parallel and then spliced together by the remote execution service. This means that we need to materialize the full file bytes, so using a piped stream wouldn't be helpful unless we changed the upload technique.